container/acl/acl_basic.go

```diff
@@ -0,0 +1,276 @@
+		OpObjectHead,
+		OpObjectPut,
+		OpObjectSearch,
+		OpObjectHash:
+		return true
+	}
+}
+
+// AllowOp allows the parties with the given role to the given operation.
+// Op MUST be one of the Op enumeration. Role MUST be one of:
+//  RoleOwner
+//  RoleContainer
+//  RoleOthers
+// and if role is RoleContainer, op MUST NOT be:
+//  OpObjectGet
+//  OpObjectHead
+//  OpObjectPut
+//  OpObjectSearch
+//  OpObjectHash
+//
+// See also IsOpAllowed.
+func (x *Basic) AllowOp(op Op, role Role) {
+	var bitPos uint8
+
+	switch role {
+	default:
+		panic(fmt.Sprintf("unable to set rules for unsupported role %v", role))
+	case RoleInnerRing:
+		panic("basic ACL MUST NOT be modified for Inner Ring")
+	case RoleOwner:
+		bitPos = opBitPosOwner
+	case RoleContainer:
+		if isReplicationOp(op) {
+			panic("basic ACL for container replication ops MUST NOT be modified")
+		}
+
+		bitPos = opBitPosContainer
+	case RoleOthers:
+		bitPos = opBitPosOthers
+	}
+
+	setOpBit((*uint32)(x), op, bitPos)
+}
+
+// IsOpAllowed checks if parties with the given role are allowed to the given op
+// according to the Basic rules. Op MUST be one of the Op enumeration.
+// Role MUST be one of the Role enumeration.
+//
+// Members with RoleContainer role have exclusive default access to the
+// operations of the data replication mechanism:
+//  OpObjectGet
+//  OpObjectHead
+//  OpObjectPut
+//  OpObjectSearch
+//  OpObjectHash
+//
+// RoleInnerRing members are allowed to data audit ops only:
+//  OpObjectGet
+//  OpObjectHead
+//  OpObjectHash
+//  OpObjectSearch
+//
+// Zero Basic prevents any role from accessing any operation in the absence
+// of default rights.
+//
+// See also AllowOp.
+func (x Basic) IsOpAllowed(op Op, role Role) bool {
+	var bitPos uint8
+
+	switch role {
+	default:
+		panic(fmt.Sprintf("role is unsupported %v", role))
+	case RoleInnerRing:
+		switch op {
+		case
+			OpObjectGet,
+			OpObjectHead,
+			OpObjectHash,
+			OpObjectSearch:
+			return true
+		default:
+			return false
+		}
+	case RoleOwner:
+		bitPos = opBitPosOwner
+	case RoleContainer:
+		if isReplicationOp(op) {
+			return true
+		}
+
+		bitPos = opBitPosContainer
+	case RoleOthers:
+		bitPos = opBitPosOthers
+	}
+
+	return isOpBitSet(uint32(x), op, bitPos)
+}
+
+// AllowBearerRules allows bearer to provide extended ACL rules for the given
+// operation. Bearer rules doesn't depend on container ACL extensibility.
+//
+// See also AllowedBearerRules.
+func (x *Basic) AllowBearerRules(op Op) {
+	setOpBit((*uint32)(x), op, opBitPosBearer)
+}
+
+// AllowedBearerRules checks if bearer rules are allowed using AllowBearerRules.
+// Op MUST be one of the Op enumeration.
+//
+// Zero Basic disallows bearer rules for any op.
+func (x Basic) AllowedBearerRules(op Op) bool {
+	return isOpBitSet(uint32(x), op, opBitPosBearer)
+}
+
+// EncodeToString encodes Basic into hexadecimal string.
+//
+// See also DecodeString.
+func (x Basic) EncodeToString() string {
+	return strconv.FormatUint(uint64(x), 16)
```

Comment:
@cthulhu-rider with these changes we are now printing basic ACL rules without 0x prefix (e.g. in `neofs-cli get container` command):
OLD BEHAVIOR: `basic ACL: 0x1c8c8ccc (private)`
NEW BEHAVIOR: `basic ACL: 1c8c8ccc (private)`

Is this intended change?

Reply:
Yes, basic ACL string format is not strictly specified. According to docs `EncodeToString` returns hex string, and I'd expect `hex.DecodeString` to not return an `error` on it (it does with `0x` prefix). @vdomnich-yadro do you expect in `neofs-cli` to get output with `0x`? If so, it should be resolved on CLI side.

Reply:
@cthulhu-rider we have test that verifies that "container get" returns correct basic ACL. The test was relying on `0x` prefix and it fails when it runs with the new CLI binaries.

I will update the test to work with the new ACL format (without 0x prefix). But before doing so I wanted to confirm that prefix removal was intended and that formatting won't change again in nearby future.
Thank you!